storage/inkless/src/test/java/io/aiven/inkless/consume/ReaderTest.java

```diff
@@ -92,4 +100,108 @@ public void testClose() throws Exception {
+            try {
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            return extentC;
+        });
+
+        // Create the ordered list: A, B, C
+        final List<CompletableFuture<FileExtent>> orderedFutures = List.of(futureA, futureB, futureC);
+
+        // Call allOfFileExtents and wait for result
+        final CompletableFuture<List<FileExtent>> resultFuture = Reader.allOfFileExtents(orderedFutures);
+        final List<FileExtent> result = resultFuture.get(5, TimeUnit.SECONDS);
+
+        // Verify result order is preserved as A, B, C (not C, B, A which was the completion order)
+        assertThat(result)
+            .hasSize(3)
+            .extracting(FileExtent::object)
+            .containsExactly(objectKeyA.value(), objectKeyB.value(), objectKeyC.value());
+    }
+
+    /**
+     * Tests that allOfFileExtents returns immediately without blocking the calling thread.
+     */
+    @Test
+    public void testAllOfFileExtentsDoesNotBlock() {
+        // Create file extents
+        final ObjectKey objectKey = PlainObjectKey.create("prefix", "object");
+        final FileExtent extent = FileFetchJob.createFileExtent(objectKey, new ByteRange(0, 10), ByteBuffer.allocate(10));
+
+        // Create a future that will complete after a delay
+        final AtomicBoolean futureCompleted = new AtomicBoolean(false);
+        final CompletableFuture<FileExtent> delayedFuture = CompletableFuture.supplyAsync(() -> {
+            try {
+                Thread.sleep(1000); // Simulate slow operation
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+            futureCompleted.set(true);
+            return extent;
+        });
+
+        final List<CompletableFuture<FileExtent>> futures = List.of(delayedFuture);
+
+        // Call allOfFileExtents - this should return immediately
+        final long startTime = System.currentTimeMillis();
+        final CompletableFuture<List<FileExtent>> resultFuture = Reader.allOfFileExtents(futures);
+        final long callDuration = System.currentTimeMillis() - startTime;
+
+        // Verify the call returned immediately (within 100ms)
+        assertThat(callDuration).isLessThan(100);
+
+        // Verify the future is not yet complete
+        assertThat(resultFuture).isNotCompleted();
+        assertThat(futureCompleted).isFalse();
+
+        // Wait for the result to actually complete
+        await().atMost(2, TimeUnit.SECONDS).untilAsserted(() -> {
+            assertThat(resultFuture).isCompleted();
+            assertThat(futureCompleted).isTrue();
+        });
+
+        // Verify the result is correct
+        assertThat(resultFuture.join())
+            .hasSize(1)
+            .extracting(FileExtent::object)
+            .containsExactly(objectKey.value());
```

Comment:
Here again I think we are relying on `Thread.sleep` and call durations too much but we could just manually complete the futures exactly when we need. Something like this for example
```
        final CompletableFuture<FileExtent> incompleteFuture = new CompletableFuture<>();

        final CompletableFuture<List<FileExtent>> resultFuture = Reader.allOfFileExtents(List.of(incompleteFuture));

        // Proves non-blocking: method returned but result not ready
        assertThat(resultFuture).isNotCompleted();

        // Complete input
        final FileExtent extent = createExtent("test");
        incompleteFuture.complete(extent);

        // Verify completion and basic correctness
        assertThat(resultFuture.join()).containsExactly(extent);
```